homeassistant/components/asuswrt/config_flow.py

```diff
@@ -0,0 +1,175 @@
+    try:
+        return socket.gethostbyname(host)
+    except socket.gaierror:
+        return None
+
+
+class AsusWrtFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize AsusWrt config flow."""
+        self._host = None
+        self._name = None
+
+    def _show_setup_form(self, user_input=None, errors=None):
+        """Show the setup form to the user."""
+
+        if user_input is None:
+            user_input = {}
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_HOST, default=user_input.get(CONF_HOST, "")): str,
+                    vol.Optional(CONF_NAME, default=user_input.get(CONF_NAME, "")): str,
+                    vol.Required(
+                        CONF_USERNAME, default=user_input.get(CONF_USERNAME, "")
+                    ): str,
+                    vol.Optional(CONF_PASSWORD): str,
+                    vol.Optional(CONF_SSH_KEY): str,
+                    vol.Required(CONF_PROTOCOL, default="ssh"): vol.In(
+                        {"ssh": "SSH", "telnet": "Telnet"}
+                    ),
+                    vol.Required(CONF_PORT, default=DEFAULT_SSH_PORT): cv.port,
+                    vol.Required(CONF_MODE, default="router"): vol.In(
+                        {"router": "Router", "ap": "Access Point"}
+                    ),
+                    vol.Required(CONF_REQUIRE_IP, default=True): bool,
+                    vol.Required(CONF_INTERFACE, default=DEFAULT_INTERFACE): str,
+                    vol.Required(CONF_DNSMASQ, default=DEFAULT_DNSMASQ): str,
+                }
+            ),
+            errors=errors or {},
+        )
+
+    async def _async_check_connection(self, user_input):
+        """Attempt to connect the AsusWrt router."""
+
+        errors = {}
+
+        api = get_api(user_input)
+        try:
+            await api.connection.async_connect()
+            if api.is_connected:
+                if hasattr(api.connection, "disconnect"):
+                    await api.connection.disconnect()
+
+                return self.async_create_entry(
+                    title=self._name,
+                    data=user_input,
+                )
+
+            _LOGGER.error("Error connecting to the AsusWrt router at %s", self._host)
+            errors["base"] = "cannot_connect"
+
+        except OSError:
+            _LOGGER.error("Error connecting to the AsusWrt router at %s", self._host)
+            errors["base"] = "cannot_connect"
+
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception(
+                "Unknown error connecting with AsusWrt router at %s", self._host
+            )
+            errors["base"] = "unknown"
+
+        return self._show_setup_form(user_input, errors)
```

Comment:
Just check the connection and return the validated data on success and raise an error on error. Don't create the entry and don't show the form from this helper.

Reply:
See our scaffold script for an example.